File: AiSD1/L1/go/lifo.go
package main

import "fmt"

type Element struct {
	value int
	next  *Element
}

type Stack struct {
	top  *Element
	size int
}

func createStack() *Stack {
	return &Stack{top: nil, size: 0}
}

func isEmpty(s *Stack) bool {
	return s.size == 0
}

func initStack(s *Stack, val int) {
	e := &Element{value: val, next: nil}
	s.top = e
	s.size = 1
}

func push(s *Stack, val int) {

	if isEmpty(s) {
		initStack(s, val)
		return
	}

	e := &Element{value: val, next: s.top}
	s.top = e
	s.size++
}

func pop(s *Stack) bool {
	if isEmpty(s) {
		fmt.Println("Stos pusty!")
		return false
	}

	tmp := s.top.next
	s.top = tmp 
	s.size--
	return true
}

func print(s *Stack) {
	if isEmpty(s) {
		fmt.Println("Stack is empty!!!")
	} else {
		fmt.Println(" ")
		fmt.Println("--------------------")
		fmt.Println(" ")
		curr := s.top
		for curr != nil {
			fmt.Println(curr.value)
			fmt.Println("↓")
			curr = curr.next
		}
		fmt.Println("NIL")
		fmt.Println(" ")
		fmt.Println("--------------------")
		fmt.Println(" ")

	}

}


func main() {
	stack := createStack()
	pop(stack)
	push(stack, 3)
	push(stack, 13)
	push(stack, 21)
	push(stack, 37)
	push(stack, 2115)
	print(stack)
	pop(stack)
	print(stack)
}

File: AiSD1/L1/lifo.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

typedef struct Element Element;
typedef struct Stack Stack;

struct Element {
    int value;
    Element* next;
};

struct Stack {
    Element* top;
    int size;
};

Stack* createStack() {
    Stack* stack = malloc (sizeof (Stack));
    stack -> top = NULL;
    stack -> size = 0;
    return stack;
}

bool isEmpty(Stack* stack) {
    if (stack->size == 0) {
        return true;
    }
    return false;
}

void initStack(Stack* stack, int value) {
    Element* element = malloc (sizeof(Element));
    element->value = value;
    element->next = NULL;
    stack->top = element;
    stack->size = 1;
}

void push(Stack* stack, int value) {

    if(isEmpty(stack)) {
        initStack(stack, value);
    } else {
        Element *element = malloc(sizeof(Element));
        element->value = value;
        element->next = stack->top;
        stack->top = element;
        stack->size++;
    }
 
}

bool pop(Stack* stack) {
 
    if (isEmpty(stack)) {
        printf("EMPTY!\n");
        return false;
	} else {
        Element* tmp = stack->top->next;
        free(stack->top);
        stack->top = tmp;
        stack->size--;
        return true;
	}
 
}

void show(Stack* stack) {
    if (isEmpty(stack)) {
         printf("Stack is empty\n");
    }
    else {
        Element* current = stack->top;
        while (current != NULL) {
            printf("%i", current->value);
            printf("\n");
            printf("↓");
            printf("\n");
            current = current->next;
         }
    }
}



int main() {
        Stack* stack = createStack();
        pop(stack);
        push(stack, 3);
        push(stack, 13);
        push(stack, 21);
        push(stack, 37);
        push(stack, 2115);
        show(stack);
        pop(stack);
        printf("\n");
        show(stack);
}